PayloadPkg/FirmwareUpdate/FirmwareUpdateHelper.c

```diff
@@ -1091,6 +1091,80 @@ UpdateSblComponent (
   return Status;
 }
 
+/**
+  Read the value of FW_UPDATE_STATUS.CsmeNeedReset
+
+  The CsmeNeedReset flag is used to ensure CSME update
+  has taken effect before processing CMDI payload.
+  This is specific to prevent {OEMKEYREVOCATION} command
+  failure for the case that CSME payload contains OEM KM
+  with key revocation extension.
+
+  @param[out] Value  Value of FW_UPDATE_STATUS.CsmeNeedReset
+
+  @retval  Value  Value of FW_UPDATE_STATUS.CsmeNeedReset
+**/
+UINT8
+ReadCsmeNeedResetFlag (
+  VOID
+  )
+{
+  EFI_STATUS  Status;
+  UINT32      FwUpdStatusOffset;
+  UINT8       Value;
+
+  FwUpdStatusOffset = PcdGet32(PcdFwUpdStatusBase);
+  FwUpdStatusOffset += OFFSET_OF(FW_UPDATE_STATUS, CsmeNeedReset);
+
+  Value = CSME_NEED_RESET_INIT;
+  Status = BootMediaRead (FwUpdStatusOffset, sizeof(UINT8), (UINT8 *)&Value);
+  if (EFI_ERROR (Status)) {
+    DEBUG((DEBUG_ERROR, "BootMediaRead CsmeNeedReset. offset: 0x%04x, Status = 0x%x\n",
+          FwUpdStatusOffset, Status));
+    Value = CSME_NEED_RESET_INVALID;
+  }
+
+  return Value;
+}
+
+/**
+  Write the value of FW_UPDATE_STATUS.CsmeNeedReset
+
+  @param[in] Value  Value to be written to FW_UPDATE_STATUS.CsmeNeedReset
+
+  @retval  EFI_SUCCESS            Write operation is successful
+  @retval  EFI_INVALID_PARAMETER  Invalid parameter
+  @retval  EFI_DEVICE_ERROR       Write operation failed
+**/
+EFI_STATUS
+WriteCsmeNeedResetFlag (
+  IN  UINT8  Value
+  )
+{
+  EFI_STATUS  Status;
+  UINT32      FwUpdStatusOffset;
+  UINT8       CurrVal;
+
+  CurrVal = ReadCsmeNeedResetFlag();
+  if (Value > CurrVal) {
```

Comment:
because CSME_NEED_RESET_<state> only uses the lower 4 bits of FW_UPDATE_STATUS.CsmeNeedReset, it implies the high 4 bits are reserved for future. Thus, it is better to either use a MASK before comparison or redefine the  FW_UPDATE_STATUS.CsmeNeedReset as a bit-field member. Similarly, when update the field, also need to reserve current high 4 bits. 

Note: it is also impossible to reduce the member to 3-bit width for 4 states: 111, 110, 100, 000.

Reply:
Thanks for figuring out it. Simply re-defined CSME_NEED_RESET_$state to use 8-bit values (0xFF, FE, FC, F8) to accommodate it.